### PR TITLE
Added private flag toggle function to ethtool

### PR DESCRIPTION
**Pull request by manvanthar**
Added private flag toggle function to ethtool

Signed-off-by: Manvanthara Puttashankar <manvanth@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by manvanthar** [comment]
Run logs for privflags not available

avocado run ethtool_test.py -m ethtool_test.py.data/ethtool_infiniband.yaml 
JOB ID     : cb574aea729efb3af0f49bb7825e581489763588
JOB LOG    : /root/tests/results/job-2019-11-27T07.20-cb574ae/job.log
 (01/16) ethtool_test.py:Ethtool.test_ethtool;arg-74b7: PASS (5.57 s)
 (02/16) ethtool_test.py:Ethtool.test_ethtool;Driver-492c: PASS (5.65 s)
 (03/16) ethtool_test.py:Ethtool.test_ethtool;Coalescing-212b: PASS (5.53 s)
 (04/16) ethtool_test.py:Ethtool.test_ethtool;RingParameter-a938: PASS (5.53 s)
 (05/16) ethtool_test.py:Ethtool.test_ethtool;ShowOffloadProtocolFeature-5ea9: PASS (5.66 s)
 (06/16) ethtool_test.py:Ethtool.test_ethtool;Statistics-4a09: PASS (6.00 s)
 (07/16) ethtool_test.py:Ethtool.test_ethtool;Show-permaddr-0a2e: PASS (5.63 s)
 (08/16) ethtool_test.py:Ethtool.test_ethtool;Show-time-stamping-91d1: PASS (5.66 s)
 (09/16) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;arg-74b7: CANCEL: Device Doesn't support Private flags (5.08 s)
 (10/16) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;Driver-492c: CANCEL: Device Doesn't support Private flags (5.02 s)
 (11/16) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;Coalescing-212b: CANCEL: Device Doesn't support Private flags (5.10 s)
 (12/16) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;RingParameter-a938: CANCEL: Device Doesn't support Private flags (5.09 s)
 (13/16) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;ShowOffloadProtocolFeature-5ea9: CANCEL: Device Doesn't support Private flags (5.02 s)
 (14/16) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;Statistics-4a09: CANCEL: Device Doesn't support Private flags (5.09 s)
 (15/16) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;Show-permaddr-0a2e: CANCEL: Device Doesn't support Private flags (5.01 s)
 (16/16) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;Show-time-stamping-91d1: CANCEL: Device Doesn't support Private flags (6.02 s)
RESULTS    : PASS 8 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 8
JOB TIME   : 87.12 s
JOB HTML   : /root/tests/results/job-2019-11-27T07.20-cb574ae/results.html

**Comment by manvanthar** [comment]
Run log for cases where there are multiple priv flags.   script makes an attempt to  toggle all, lists only which fails to be changed.

avocado run ethtool_test.py -m ethtool_test.py.data/ethtool_test.yaml
Failed to load plugin from module "avocado_runner_vm": DistributionNotFound(Requirement.parse('libvirt-python'), None) :
  File "/usr/local/lib/python3.6/site-packages/avocado_framework-72.0-py3.6.egg/avocado/core/extension_manager.py", line 62, in __init__
    plugin = ep.load()
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2317, in load
    self.require(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2340, in require
    items = working_set.resolve(reqs, env, installer, extras=self.extras)
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 774, in resolve
    raise DistributionNotFound(req, requirers)

JOB ID     : 2d2e8c306d4d1783869293ad1519f1427f556d75
JOB LOG    : /root/avocado-fvt-wrapper/results/job-2019-11-27T20.02-2d2e8c3/job.log
 (01/32) ethtool_test.py:Ethtool.test_ethtool;arg-6455: PASS (5.78 s)
 (02/32) ethtool_test.py:Ethtool.test_ethtool;Driver-68df: PASS (5.63 s)
 (03/32) ethtool_test.py:Ethtool.test_ethtool;Autonegotiation-917a: FAIL: failed (5.41 s)
 (04/32) ethtool_test.py:Ethtool.test_ethtool;PauseParameter-6ada: PASS (5.65 s)
 (05/32) ethtool_test.py:Ethtool.test_ethtool;Coalescing-a85c: PASS (5.70 s)
 (06/32) ethtool_test.py:Ethtool.test_ethtool;RegisterDump-3af7: FAIL: failed (5.45 s)
 (07/32) ethtool_test.py:Ethtool.test_ethtool;EEPROMDump-8a0d: FAIL: failed (5.40 s)
 (08/32) ethtool_test.py:Ethtool.test_ethtool;RingParameter-4d8e: PASS (5.66 s)
 (09/32) ethtool_test.py:Ethtool.test_ethtool;ShowOffloadProtocolFeature-2721: PASS (5.64 s)
 (10/32) ethtool_test.py:Ethtool.test_ethtool;AdapterInitiate-20bc: FAIL: failed (5.41 s)
 (11/32) ethtool_test.py:Ethtool.test_ethtool;Statistics-da34: PASS (5.68 s)
 (12/32) ethtool_test.py:Ethtool.test_ethtool;SelfTest-2c74: FAIL: failed (5.39 s)
 (13/32) ethtool_test.py:Ethtool.test_ethtool;Show-permaddr-c4d9: PASS (5.61 s)
 (14/32) ethtool_test.py:Ethtool.test_ethtool;Get-dump-7360: PASS (5.67 s)
 (15/32) ethtool_test.py:Ethtool.test_ethtool;Show-time-stamping-8d2b: PASS (5.66 s)
 (16/32) ethtool_test.py:Ethtool.test_ethtool;Get-phy-tunable-142e: FAIL: failed (5.38 s)
 (17/32) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;arg-6455: FAIL: Private flags could not be toggled ['xdp_tx_mpwqe', 'dropless_rq'] (9.94 s)
 (18/32) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;Driver-68df: FAIL: Private flags could not be toggled ['xdp_tx_mpwqe', 'dropless_rq'] (10.05 s)
 (19/32) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;Autonegotiation-917a: FAIL: Private flags could not be toggled ['xdp_tx_mpwqe', 'dropless_rq'] (10.01 s)
 (20/32) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;PauseParameter-6ada: FAIL: Private flags could not be toggled ['xdp_tx_mpwqe', 'dropless_rq'] (9.76 s)
 (21/32) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;Coalescing-a85c: FAIL: Private flags could not be toggled ['xdp_tx_mpwqe', 'dropless_rq'] (9.79 s)
 (22/32) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;RegisterDump-3af7: FAIL: Private flags could not be toggled ['xdp_tx_mpwqe', 'dropless_rq'] (9.63 s)
 (23/32) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;EEPROMDump-8a0d: FAIL: Private flags could not be toggled ['xdp_tx_mpwqe', 'dropless_rq'] (10.39 s)
 (24/32) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;RingParameter-4d8e: FAIL: Private flags could not be toggled ['xdp_tx_mpwqe', 'dropless_rq'] (9.71 s)
 (25/32) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;ShowOffloadProtocolFeature-2721: FAIL: Private flags could not be toggled ['xdp_tx_mpwqe', 'dropless_rq'] (10.11 s)
 (26/32) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;AdapterInitiate-20bc: FAIL: Private flags could not be toggled ['xdp_tx_mpwqe', 'dropless_rq'] (10.18 s)
 (27/32) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;Statistics-da34: FAIL: Private flags could not be toggled ['xdp_tx_mpwqe', 'dropless_rq'] (10.06 s)
 (28/32) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;SelfTest-2c74: FAIL: Private flags could not be toggled ['xdp_tx_mpwqe', 'dropless_rq'] (9.51 s)
 (29/32) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;Show-permaddr-c4d9: FAIL: Private flags could not be toggled ['xdp_tx_mpwqe', 'dropless_rq'] (10.15 s)
 (30/32) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;Get-dump-7360: FAIL: Private flags could not be toggled ['xdp_tx_mpwqe', 'dropless_rq'] (9.47 s)
 (31/32) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;Show-time-stamping-8d2b: FAIL: Private flags could not be toggled ['xdp_tx_mpwqe', 'dropless_rq'] (10.12 s)
 (32/32) ethtool_test.py:Ethtool.test_ethtool_toggle_priv_flags;Get-phy-tunable-142e: FAIL: Private flags could not be toggled ['xdp_tx_mpwqe', 'dropless_rq'] (10.38 s)
RESULTS    : PASS 10 | ERROR 0 | FAIL 22 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 249.45 s
JOB HTML   : /root/avocado-fvt-wrapper/results/job-2019-11-27T20.02-2d2e8c3/results.html

**Comment by manvanthar** [comment]
[job.log](https://github.com/avocado-framework-tests/avocado-misc-tests/files/3897311/job.log)

File: io/net/ethtool_test.py [hunk 1]
```diff
@@ -102,19 +110,41 @@ def test_ethtool(self):
...
+            if "flags" not in line:
+                priv_flag = line.split(':')[0]
+                cmd = "ethtool --set-priv-flags %s \"%s\" %s" % \
+                      (self.iface, priv_flag.rstrip(), val)
+                ret1 = process.run(cmd, shell=True, verbose=True,
+                                   ignore_status=True)
+                if ret1.exit_status:
+                    priv_fail.append(priv_flag.rstrip())
+                else:
+                    priv_pass.append(priv_flag.rstrip())
```
Comment (narasimhan-v) [inline io/net/ethtool_test.py]: @manvanthar I assume what you want to do is:
```
for every private flag:
    store current value - on/off
    try toggle the value.
    if pass:
        toggle again - to restore to original state
    if fails:
        append it to fail_list
        continue
if fail_list:
    print list of failures
```

Comment (narasimhan-v) [inline io/net/ethtool_test.py]: This is not needed

Comment (manvanthar) [reply io/net/ethtool_test.py]: @narasimhan-v  eventhough we dont explicitly print this.   in job.log actually gives us the pass and fail output like below.   Usefull to get the list of what priv_flags could be set/unset basically passed and failed.

2019-11-27 20:04:18,968 test             L0869 DEBUG|  -> cmd <class 'str'>: ethtool --set-priv-flags eth3 "tx_xdp_hw_checksum" off
2019-11-27 20:04:18,968 test             L0869 DEBUG|  -> priv_fail <class 'list'>: ['xdp_tx_mpwqe', 'dropless_rq']
2019-11-27 20:04:18,968 test             L0869 DEBUG|  -> priv_pass <class 'list'>: ['rx_cqe_moder', 'tx_cqe_moder', 'rx_cqe_compress', 'rx_striding_rq', 'rx_no_csum_complete', 'sniffer', 'per_channel_stats', 'tx_xdp_hw_checksum']

File: io/net/ethtool_test.py [hunk 2]
```diff
@@ -102,19 +110,41 @@ def test_ethtool(self):
...
-                          ignore_status=True) != 0:
-            return False
-        return True
+        priv_pass = []
+        priv_fail = []
+        cmd = "ethtool --show-priv-flags %s" % (self.iface)
+        ret = process.run(cmd, shell=True, verbose=True,
+                          ignore_status=True)
+        if ret.exit_status:
+            self.cancel("Device Doesn't support Private flags")
```
[the remark this answers is not in the document]
Comment (manvanthar) [reply io/net/ethtool_test.py]: @narasimhan-v   
we have 2 tests for ethtool.
Generic_ethtool test
ethttool_priv_flags

moving it to Setup will break the Generic_ethtool test

File: io/net/ethtool_test.py (hunk 1 above)
Comment (manvanthar) [reply io/net/ethtool_test.py]: @narasimhan-v 

Here is the sampleout of show-priv-flags

ethtool --show-priv-flags eth4
Private flags for eth4:
rx_cqe_moder       : off
tx_cqe_moder       : off
rx_cqe_compress    : off
rx_striding_rq     : off
rx_no_csum_complete: off
xdp_tx_mpwqe       : off
sniffer            : off
dropless_rq        : off
per_channel_stats  : on
tx_xdp_hw_checksum : off

Logic is
1. ignore first line and toggle all the priv flags for the device
2. make a list of pass and failed priv flags
3. Ping_check with the overall toggled list.
4. Print the fail list.

Comment (narasimhan-v) [reply io/net/ethtool_test.py]: @manvanthar Ah ok, agreed.

Comment (narasimhan-v) [reply io/net/ethtool_test.py]: @manvanthar I get this. Shouldn't the test test toggling back ?
Also, shouldn't the test bring the flags back to original state ?

File: io/net/ethtool_test.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (narasimhan-v) [reply io/net/ethtool_test.py]: @manvanthar str(self.name.name) will have test name, which can be used in `setUp()`. This additionally can be used along with your logic to cancel

File: io/net/ethtool_test.py [hunk 3]
```diff
@@ -102,19 +110,41 @@ def test_ethtool(self):
...
+        for line in ret.stdout.decode("utf-8").splitlines():
```
[the remark this answers is not in the document]
Comment (manvanthar) [reply io/net/ethtool_test.py]: ok Can be done

File: io/net/ethtool_test.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (manvanthar) [reply io/net/ethtool_test.py]: ok done

**Comment by manvanthar** [comment]
review comments addressed in PR#1467

**Pull request closed without merging**